### PR TITLE
add an option to start the relay v2

### PR DESCRIPTION
**Pull request by marten-seemann**
I'm not satisfied with the names for the config options, but I struggle to come up with something better. Suggestions would be highly appreciated.

I wish we could use `EnableRelay()` here, but unfortunately this is already taken to enable the *relay transport*.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 	"github.com/libp2p/go-libp2p/p2p/host/relay"
 	routed "github.com/libp2p/go-libp2p/p2p/host/routed"
 	circuitv2 "github.com/libp2p/go-libp2p/p2p/protocol/circuitv2/client"
+	v2relay "github.com/libp2p/go-libp2p/p2p/protocol/circuitv2/relay"
```
Comment (vyzo) [inline config/config.go:24]: let's call this `relayv2` for consistency.

File: config/config.go [hunk 2]
```diff
@@ -74,7 +75,10 @@ type Config struct {
 	PSK                pnet.PSK
 
 	RelayCustom bool
-	Relay       bool
+	Relay       bool // should the relay transport be used
+
+	DisableRelay bool // should we run a circuitv2 relay (if publicly reachable)
+	RelayOpts    []v2relay.Option
```
Comment (vyzo) [inline config/config.go:81]: `RelayService` sounds like a better name, less ambiguous.

File: config/config.go [hunk 3]
```diff
@@ -196,6 +200,8 @@ func (cfg *Config) NewNode() (host.Host, error) {
 		MultiaddrResolver:   cfg.MultiaddrResolver,
 		EnableHolePunching:  cfg.EnableHolePunching,
 		HolePunchingOptions: cfg.HolePunchingOptions,
+		EnableV2Relay:       !cfg.DisableRelay,
+		RelayOpts:           cfg.RelayOpts,
```
Comment (vyzo) [inline config/config.go:204]: `EnableRelayService` and `RelayServiceOptions` perhaps.

File: options.go [hunk 4]
```diff
@@ -19,7 +19,8 @@ import (
 	"github.com/libp2p/go-libp2p/config"
 	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
 	autorelay "github.com/libp2p/go-libp2p/p2p/host/relay"
-	holepunch "github.com/libp2p/go-libp2p/p2p/protocol/holepunch"
+	v2relay "github.com/libp2p/go-libp2p/p2p/protocol/circuitv2/relay"
```
Comment (vyzo) [inline options.go:22]: `relayv2`

File: options.go [hunk 5]
```diff
@@ -231,6 +232,26 @@ func DisableRelay() Option {
 	}
 }
 
+// DisableRelaying configures libp2p to not run a circuit v2 relay.
+// Note that we only start the relay if we detect that we're publicly reachable.
+func DisableRelaying() Option {
```
Comment (vyzo) [inline options.go:237]: `DisalbeRelayService`

File: options.go [hunk 6]
```diff
@@ -231,6 +232,26 @@ func DisableRelay() Option {
...
+	return func(cfg *Config) error {
+		cfg.DisableRelay = true
+		cfg.RelayOpts = nil
+		return nil
+	}
+}
+
+// EnableRelaying configures libp2p to run a circuit v2 relay,
+// if we dected that we're publicly reachable.
+func EnableRelaying(opts ...v2relay.Option) Option {
```
Comment (vyzo) [inline options.go:247]: `EnableRelayService`

File: p2p/host/basic/basic_host.go [hunk 7]
```diff
@@ -133,6 +137,11 @@ type HostOpts struct {
 	// EnablePing indicates whether to instantiate the ping service
 	EnablePing bool
 
+	// EnableV2Relay enables the circuit v2 relay (if we're publicly reachable).
+	EnableV2Relay bool
+	// RelayOpts are options for the circuit v2 relay.
+	RelayOpts []v2relay.Option
```
Comment (vyzo) [inline p2p/host/basic/basic_host.go:143]: `EnableRelayService` and `RelayServiceOpts`.

**Review by vyzo — COMMENTED** [review]
Yeah, these names are very confusing.

I have suggested some alternatives.

**Comment by marten-seemann** [comment]
`RelayService` sounds good to me! Will update the PR.

File: config/config.go [hunk 8]
```diff
@@ -74,7 +75,10 @@ type Config struct {
 	PSK                pnet.PSK
 
 	RelayCustom bool
-	Relay       bool
+	Relay       bool // should the relay transport be used
+
+	DisableRelayService bool // should we run a circuitv2 relay (if publicly reachable)
```
Comment (vyzo) [inline config/config.go:80]: let's heed to @stebalien's request to not enable it by default, so this should be `EnableRelayService` I guess and the logic inverted.

File: options.go [hunk 9]
```diff
@@ -231,6 +232,26 @@ func DisableRelay() Option {
...
+// DisableRelayService configures libp2p to not run a circuit v2 relay.
+// Note that we only start the relay if we detect that we're publicly reachable.
+func DisableRelayService() Option {
+	return func(cfg *Config) error {
+		cfg.DisableRelayService = true
+		cfg.RelayServiceOpts = nil
+		return nil
+	}
+}
+
```
Comment (vyzo) [inline options.go:244]: we don't need this if it is not enabled by default.

File: p2p/host/relayv2/relay.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package relayv2
```
Comment (vyzo) [inline p2p/host/relayv2/relay.go:1]: let's call this package `relaysvc` or something similar? We are moving the autorelay stuff in `autorelay` anyway.

**Review by vyzo — COMMENTED** [review]
LGTM; only thing we need is to make it disabled by default as @stebalien requested.
Also, the package name for the manager/service is rather confusing, let's call it something more appropriate.

File: config/config.go (hunk 8 above)
Comment (marten-seemann) [reply config/config.go]: Makes sense.
We'll still enable this by default in IPFS / Filecoin, won't we?

Comment (Stebalien) [reply config/config.go]: >  We'll still enable this by default in IPFS / Filecoin, won't we?

Yes, or at least in IPFS. But libp2p is a library not an application.

**Review by vyzo — APPROVED** [review]

**Pull request merged**